### PR TITLE
(v2) Use KeyMsg/MouseMsg interfaces

### PR DESCRIPTION
**Pull request by aymanbagabas**
This replaces KeyMsg and MouseMsg with interfaces that catch their respective types. For example, both KeyPressMsg and KeyReleaseMsg implement KeyMsg. Same goes for MouseClickMsg, MouseReleaseMsg, etc, they all implement MouseMsg. This makes it possible to switch on the interface type to catch all events of the same input message type.

New API:

```go

// KeyMsg represents a key event. This can be either a key press or a key
// release event.
type KeyMsg interface {
	fmt.Stringer

	// Key returns the underlying key event.
	Key() Key
}


// Key represents a Key press or release event. It contains information about
// the Key pressed, like the runes, the type of Key, and the modifiers pressed.
// There are a couple general patterns you could use to check for key presses
// or releases:
//
//	// Switch on the string representation of the key (shorter)
//	switch msg := msg.(type) {
//	case KeyPressMsg:
//	    switch msg.String() {
//	    case "enter":
//	        fmt.Println("you pressed enter!")
//	    case "a":
//	        fmt.Println("you pressed a!")
//	    }
//	}
//
//	// Switch on the key type (more foolproof)
//	switch msg := msg.(type) {
//	case KeyMsg:
//	    // catch both KeyPressMsg and KeyReleaseMsg
//	    switch key := msg.Key(); key.Code {
//	    case KeyEnter:
//	        fmt.Println("you pressed enter!")
//	    default:
//	        switch key.Text {
//	        case "a":
//	            fmt.Println("you pressed a!")
//	        }
//	    }
//	}
//
// Note that [Key.Text] will be empty for special keys like [KeyEnter],
// [KeyTab], and for keys that don't represent printable characters like key
// combos with modifier keys. In other words, [Key.Text] is populated only for
// keys that represent printable characters shifted or unshifted (like 'a',
// 'A', '1', '!', etc.).
type Key struct {
	// Text contains the actual characters received. This usually the same as
	// [Key.Code]. When [Key.Text] is non-empty, it indicates that the key
	// pressed represents printable character(s).
	Text string

	// Mod represents modifier keys, like [ModCtrl], [ModAlt], and so on.
	Mod KeyMod

	// Code represents the key pressed. This is usually a special key like
	// [KeyTab], [KeyEnter], [KeyF1], or a printable character like 'a'.
	Code rune

	// ShiftedCode is the actual, shifted key pressed by the user. For example,
	// if the user presses shift+a, or caps lock is on, [Key.ShiftedCode] will
	// be 'A' and [Key.Code] will be 'a'.
	//
	// In the case of non-latin keyboards, like Arabic, [Key.ShiftedCode] is the
	// unshifted key on the keyboard.
	//
	// This is only available with the Kitty Keyboard Protocol or the Windows
	// Console API.
	ShiftedCode rune

	// BaseCode is the key pressed according to the standard PC-101 key layout.
	// On international keyboards, this is the key that would be pressed if the
	// keyboard was set to US PC-101 layout.
	//
	// For example, if the user presses 'q' on a French AZERTY keyboard,
	// [Key.BaseCode] will be 'q'.
	//
	// This is only available with the Kitty Keyboard Protocol or the Windows
	// Console API.
	BaseCode rune

	// IsRepeat indicates whether the key is being held down and sending events
	// repeatedly.
	//
	// This is only available with the Kitty Keyboard Protocol or the Windows
	// Console API.
	IsRepeat bool
}


// MouseMsg represents a mouse message. This is a generic mouse message that
// can represent any kind of mouse event.
type MouseMsg interface {
	fmt.Stringer

	// Mouse returns the underlying mouse event.
	Mouse() Mouse
}

// Mouse represents a Mouse message. Use [MouseMsg] to represent all mouse
// messages.
//
// The X and Y coordinates are zero-based, with (0,0) being the upper left
// corner of the terminal.
//
//	// Catch all mouse events
//	switch msg := msg.(type) {
//	case MouseMsg:
//	    m := msg.Mouse()
//	    fmt.Println("Mouse event:", m.X, m.Y, m)
//	}
//
//	// Only catch mouse click events
//	switch msg := msg.(type) {
//	case MouseClickMsg:
//	    fmt.Println("Mouse click event:", msg.X, msg.Y, msg)
//	}
type Mouse struct {
	X, Y   int
	Button MouseButton
	Mod    KeyMod
}

```

TODO: update examples (in a separate PR)

### CODE REVIEW

File: key.go [hunk 1]
```diff
@@ -256,54 +297,130 @@ type Key struct {
...
 
 // KeyReleaseMsg represents a key release message.
-type KeyReleaseMsg Key
+type KeyReleaseMsg key
+
+var _ KeyMsg = KeyReleaseMsg{}
 
 // String implements fmt.Stringer and is quite useful for matching complex key
-// events. For details, on what this returns see [Key.String].
+// events. For details, on what this returns see [key.String].
```
Comment (bashbunni) [inline key.go:370]: This link doesn't work in the godoc when viewed in browser

File: key.go [hunk 2]
```diff
@@ -210,29 +213,67 @@ const (
...
 	// of 1. Use [Rune()] to get the first key rune received. If the user
-	// presses shift+a, the Runes will be `[]rune{'A'}`.
-	Runes []rune
+	// presses shift+a, the runes will be `[]rune{'A'}`.
+	runes []rune
 
-	// Type is a special key, like enter, tab, backspace, and so on.
-	Type KeyType
+	// typ is a special key, like enter, tab, backspace, and so on.
+	typ KeyType
```
Comment (bashbunni) [inline key.go:276]: should this be `type`?

**Review by bashbunni — COMMENTED** [review]
There are some magic numbers in there too that the linter picked up. The changes look good though, stoked to have kitty keyboard support 🚀

File: key.go [hunk 3]
```diff
@@ -210,29 +213,67 @@ const (
 //	// Switch on the key type (more foolproof)
```
[the remark this answers is not in the document]
Comment (aymanbagabas) [reply key.go]: Now it's even simpler, `msg.Code == 'c' && msg.Mod == tea.ModCtrl`

File: key.go (hunk 2 above)
Comment (aymanbagabas) [reply key.go]: `type` is a go reserved word

File: key.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (meowgorithm) [reply key.go]: What about `msg.Type` in this example?

[the remark this answers is not in the document]
Comment (aymanbagabas) [reply key.go:248]: I've added the examples back, accidentally removed them

**Pull request merged**